### PR TITLE
Build with NO_MMAP

### PR DESCRIPTION
**Pull request by 0xdky**
This patch fixes building libgit2 without mmap (NO_MMAP)

Since 'mmap' does not update the position in file descriptor. Hence, we have to capture and restore the position of file descriptor in 'mmap' emulation.

### CODE REVIEW

**Comment by ethomson** [comment]
Cool, thanks for tackling this.  :)

Stylistically, I would slightly prefer it if we didn't interleave big blocks of `ifdef` right before the return.  It feels a little hard to reason about what the mmap emulation in `write_at` is doing.  Let's make them self-contained blocks so that the flow of the function is:

```
static int write_at(git_indexer *idx, const void *data, off64_t offset, size_t size)
{
#ifdef NO_MMAP
    ...mmap emulation goes here...
    return;
#else
    ...mmap usage goes here...
    return;
#endif
}
```

We could also just have `write_at` as an indirection between mmap-using and mmap-not-using functions, but what I don't love is the dangling completion in the `NO_MMAP` case, you have to scroll through a bunch of `ifdef`d out code to see what happens next in the function.

**Comment by hackworks** [comment]
@ethomson I have incorporated the suggested changes. Could I please get some reviews on this pull request?

File: src/posix.c
Comment (lhchavez) [inline src/posix.c:244]: `EPIPE` is not in the list of errors that can be returned by either [`read(2)`](https://man7.org/linux/man-pages/man2/read.2.html#ERRORS) or [`lseek(2)`](https://man7.org/linux/man-pages/man2/lseek.2.html#ERRORS) (which [`pread(2)`](https://man7.org/linux/man-pages/man2/pread.2.html#ERRORS) claims to be the legal errors).

instead, maybe break here like in the `p_read()` implementation. the caller would still need to check whether the return value was a short read, for consistency with `p_read()`, but that's already done.

File: src/posix.c
Comment (lhchavez) [inline src/posix.c:323]: where is `idx` declared?

File: src/posix.c
Comment (lhchavez) [inline src/posix.c:321]: this style doesn't seem to appear anywhere else in the codebase. would it be possible to use the more common

```c
if ((error = ...) < 0)
  goto cleanup;
if ((error = ...) < 0)
  goto cleanup;
...
cleanup:
  close(nfd);
  return error;
```

File: src/posix.c
Comment (hackworks) [reply src/posix.c:323]: Coding derp - copy/paste!

File: src/posix.c
Comment (lhchavez) [inline src/posix.c:229]: same as with `p_pwrite`.

**Review by lhchavez — DISMISSED** [review]
lgtm (although i am not a reviewer with write access, so my vote is just informative)

**Comment by 0xdky** [comment]
@ethomson Could I please get some help to get this merged OR if you have any concerns/review comments, I will be happy to  work on them.

File: src/indexer.c
Comment (ethomson) [inline src/indexer.c:608]: I like that you've noticed that there's a posix call that matches our intent here - our posix indirection layer is really meant to use the actual posix functions when they're available, or implement a suitable fallback when they're not.  Can we do that here?

ie, can we just use `pwrite` when it's available?  I think that it's possible that there's no system that both has `pwrite` and lacks `mmap`, but the reason that I ask isn't because I think that the posix function is necessarily super useful for us.  What it _does_ do is act as a baseline in tests to ensure that _our_ `pwrite` implementation acts the same.  This is important because we expect `p_some_posix_fn` to act like `some_posix_fn`.  If it _doesn't_ then we're going to make errors when we call it because this is really meant to be a compatibility layer (at least as much as is possible).

(If our `p_pwrite` function really _isn't_ like `pwrite` then we should just keep calling it `write_at` or something.)

Anyway, this is clever, I wouldn't have thought of emulating `pwrite`.

File: src/posix.c
Comment (ethomson) [inline src/posix.c:220]: If we have indirection between the `unix` and `win32` compatibility layers, they should go into a file in `unix/` and `win32/` directories (eg `win32/posix_w32.c`) so that we can avoid the ifdefs.

File: src/posix.c
Comment (ethomson) [inline src/posix.c:235]: `if (foo == -1)` please, not the backward form.  😄

File: src/posix.c
Comment (ethomson) [inline src/posix.c:228]: I had to read this a few times - I think that we could simplify its understandability, at the expense of an extra line or two.  

```
if (p_lseek(fd, offset, SEEK_SET) < 0) {
    error = -1;
    goto cleanup;
}
```

But since we don't really want to propagate the error value, we want to return `-1` on failure, then it might make sense to have an `on_error` label that just returns `-1`.

File: src/posix.c
Comment (ethomson) [inline src/posix.c:231]: We probably want out function to return `-1` on any failure, not propagate `p_read`'s failure value

File: src/indexer.c
Comment (ethomson) [inline src/indexer.c:607]: Also, I don't like casting a `size_t` to an `ssize_t`.  I realize that it's Probably Fine on Most Systems.  But there aren't significant guarantees about these types -- IIRC, `ssize_t` is guaranteed to support a range of `-1` to the largest chunk of data that can be written (or something equally hand waving like that).  If we take a `size_t` and call a thing that reads/writes `ssize_t`s, we should probably do it in a loop, writing as many as `SSIZE_MAX` bytes _or_ `size` at a time (whichever is smaller) until we've written all of the `size` or hit an error.

File: src/posix.c
Comment (ethomson) [inline src/posix.c:357]: Not sure that I understand this comment, is it outdated?

File: src/indexer.c
Comment (0xdky) [reply src/indexer.c:608]: I have implemented p_pwrite using pwrite or p_mmap in src/posix.c since mmap based IO was used prior to this change. If there are systems where mmap based IO is more performant, we will not cause a regression.

I will replace the call of p_pwrite in write_at using a loop to avoid the typecasting of ssize_t to size_t.

Please let me know if I am missing something.

File: src/posix.c
Comment (0xdky) [reply src/posix.c]: `p_write` in `src/posix.c` returns 0 on success or -1 on error. However, POSIX `write` returns the actual bytes written. Hence, from a consistency point of view, it would be desirable to make that change (not in the scope/related of this PR)

File: src/posix.c
Comment (0xdky) [reply src/posix.c]: Since this is win32 specific code, I moved it to `win32/posix_w32.c` and used `mmap` based IO for emulating `p_pread` and `p_pwrite`

File: src/indexer.c
Comment (0xdky) [reply src/indexer.c]: Done

**Comment by 0xdky** [comment]
@ethomson Thank you for the review and suggestions to refactor the code.

I have incorporated review comments by having unix/win32 specific POSIX emulation. On Unix, I use `pread` + `pwrite` and on win32, I use memory mapped IO since that will avoid having to reset the file descriptor position if I use `read` + `write`.

Since `mmap` itself is emulated on win32, there might be a more direct way to implement `pread` and `pwrite` using overlapped file IO.

**Comment by 0xdky** [comment]
@ethomson Build setup is failing for some environments and the results say that it has been failing since last couple of builds but with no additional logs. Could you please help me with this?

**Comment by ethomson** [comment]
> @ethomson Build setup is failing for some environments and the results say that it has been failing since last couple of builds. Could you please help me with this?

I'm deprecating the Azure Pipelines builds, moving over to GitHub Actions fully.  Please ignore the Azure Pipelines failures; apologies for the noise.

**Comment by ethomson** [comment]
If I'm reading this right, you've removed the mmap for Unix altogether?  Was this intentional?

The goal of the `p_` functions is to emulate posix like functionality.  They shouldn't wrap the posix functions and then add more.  `p_pwrite` should be able to be `#define`d directly to `pwrite` in most cases.

The looping logic should be in `write_at`

**Comment by 0xdky** [comment]
> If I'm reading this right, you've removed the mmap for Unix altogether? Was this intentional?

I just feel this really reflects my initial intent of just fixing this to support `mmap` emulation. I feel it is best to close this PR and start from fresh. Let me see if I can salvage this PR.

**Comment by 0xdky** [comment]
> If I'm reading this right, you've removed the mmap for Unix altogether? Was this intentional?
> 
> The goal of the `p_` functions is to emulate posix like functionality. They shouldn't wrap the posix functions and then add more. `p_pwrite` should be able to be `#define`d directly to `pwrite` in most cases.
> 
> The looping logic should be in `write_at`

**Comment by 0xdky** [comment]
> If I'm reading this right, you've removed the mmap for Unix altogether? Was this intentional?

I just feel this really does not reflect my initial intent of just fixing this to support mmap emulation. I feel it is best to close this PR and start from fresh. Let me see if I can salvage this PR.

> The goal of the `p_` functions is to emulate posix like functionality. They shouldn't wrap the posix functions and then add more. `p_pwrite` should be able to be `#define`d directly to `pwrite` in most cases.
> 
> The looping logic should be in `write_at`

src/posix.c has `p_read` and `p_write` that has looping to handle EINTR. I followed those as guidelines. Since we are actually making underlying `pread` and `pwrite` handle EINTR and it is not pertaining to any logic in `write_at`, I feel the EINTR based retry should still be in their `p_...` implementations.

**Comment by 0xdky** [comment]
> If I'm reading this right, you've removed the mmap for Unix altogether? Was this intentional?

Fixed it by adding conditional code in `write_at` to use `pwrite` if `NO_MMAP` is defined. Since `write_at` is not a POSIX call, I have retained conditional code in `src/indexer.c` and not moved it elsewhere.

**Comment by 0xdky** [comment]
@ethomson Could you please take a look at this updated PR? I would like to start using this functionality to build other workflows over NFS. This is a blocker when using git+libgit2 over NFS.

File: src/win32/posix_w32.c
Comment (lhchavez) [inline src/win32/posix_w32.c:994]: maybe

```suggestion
		errno = EINVAL;
```

? `E2BIG` is used for argument lists, like in [`execve(2)`](https://man7.org/linux/man-pages/man2/execve.2.html):

>        E2BIG  The total number of bytes in the environment (envp) and
>              argument list (argv) is too large.

File: src/indexer.c
Comment (0xdky) [reply src/indexer.c]: Thank you! That makes it robust.

File: src/posix.h
Comment (0xdky) [reply src/posix.h:123]: I was so used to explicitly handling EINTR, never thought GNU libc has a macro for doing that. Thank you for the above macro and the code does look cleaner now.

File: src/indexer.c
Comment (0xdky) [reply src/indexer.c]: `if (remaining_size > (size_t)(SSIZE_MAX - offset)) {
 		errno = EOVERFLOW;
 		return -1;
}`

I replaced it with the above block at the begin of function so that we catch the error early and can fail without making any changes via call to `p_pwrite`

File: src/posix.c
Comment (lhchavez) [inline src/posix.c:256]: this may need to be rewritten to call it in a loop since short reads are now possible.

File: src/win32/posix_w32.c
Comment (0xdky) [reply src/win32/posix_w32.c:1022]: Good call to let the caller drive looping.

However, I am not sure if is a good idea to mimic POSIX limits in Win32 with the following code:
```
if (!git__is_ssizet(size) || git__add_int64_overflow(&final_offset, offset, (int64_t)size)) {
	errno = EINVAL;
	return -1;	
}
```

Comment (lhchavez) [reply src/win32/posix_w32.c:1022]: given that `p_pwrite`/`p_pread` is trying to provide a POSIX-compatible implementation of `pwrite`/`pread`, it should in theory behave with the same semantics and impose similar restrictions. otherwise, the caller needs to know what semantics to expect of it, which kind of defeats the purpose of having compatibility wrappers.

(or did I misinterpret your comment?)

File: src/win32/posix_w32.c
Comment (0xdky) [reply src/win32/posix_w32.c:1022]: I incorrectly offloaded the responsibility to check for potential overflow to the caller. I agree, the underlying API implementation should try to enforce error handling as early as possible without spring a surprise when the request is retried for completion in a loop.

File: src/win32/posix_w32.c
Comment (0xdky) [reply src/win32/posix_w32.c]: I was wrong in trying to let Win32 behave differently. I agree having similar behavior is better that surprises due to differences in native implementations.

**Review by lhchavez — APPROVED** [review]
:tada:

**Comment by lhchavez** [comment]
hmm seems like I still can't merge PRs. but this is ready to go from my side!

thanks for this PR, @0xdky!

**Comment by 0xdky** [comment]
> hmm seems like I still can't merge PRs. but this is ready to go from my side!
> 
> thanks for this PR, @0xdky!

Thank you very much @lhchavez for reviewing & insightful feedback. Looking forward to @ethomson help with review+merging please.

**Comment by ethomson** [comment]
This is a _really_ crisp and clean PR.  Thanks so much for the improvements you've made to it, it really fits in nicely.  I also appreciate you going the extra mile with the Windows implementation.

I took a pass through the mmap emulation code and it looks good to me.  I didn't test it exhaustively, but this seems like it's in a very solid state for merging.

I have one or two stylistic notes, but they're more informative and not really blockers, and I want to get this in since you've been waiting on me for a while and I want to land this PR.

Thanks again for your patience and for iterating on a great PR.

File: src/indexer.c
Comment (ethomson) [inline src/indexer.c:621]: I think that I would have place the return here so that the entire function was encapsulated either in the `#if` or the `#else` blocks.  Not major, but having to jump down to see what was happening next was jarring.  It would have been nice to have each branch fully within their own space so that I didn't have to jump around.

File: src/posix.h
Comment (ethomson) [inline src/posix.h:93]: We tend to avoid macros to do things like this...  again, my preference is generally to avoid having to jump around too much within the code base to understand what a piece of code is doing.  I don't love macros in general, and in particular, I don't like macros that loop (or do flow control).

Again, there are definitely exceptions in our codebase (there's some `vector_foreach` type of macro that is in pretty regular use).  But I think that something that's used just twice is a place where we can safely repeat ourselves without too much downside.  It think that the clarity outweighs the lack of DRYness here.

This is mostly personal opinion, and again, I don't think that it's so important that we shouldn't merge this.  The benefits from this PR far outweigh any macro-related gripes that I have.  😁   Just trying to convey some stylistic thoughts.

**Review by ethomson — COMMENTED** [review]
Love this overall.  Two quick comments that might be useful for future PRs.  But overall, it looks like you really understand the libgit2 "style".  Great work.

File: src/indexer.c
Comment (0xdky) [reply src/indexer.c:621]: I agree. Having self-contained blocks makes it easier to read & you know for certain there is no dependency on anything outside the block.

Comment (0xdky) [reply src/indexer.c:621]: I agree. Having self-contained blocks makes it easier to read & you know for certain there is no dependency on anything outside the block.

**Comment by haampie** [comment]
Hi @0xdky / @ethomson, thanks for this, this is a sledgehammer approach for me to work around https://github.com/libgit2/libgit2/issues/4019.

Can we have an equivalent CMake option too? `-DLIBGIT2_NO_MMAP=ON` or so

**Pull request merged**
